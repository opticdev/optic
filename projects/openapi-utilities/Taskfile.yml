# https://taskfile.dev/#/installation
version: "3"

tasks:
  test:
    desc: Test
    cmds:
      - yarn run ws:test
