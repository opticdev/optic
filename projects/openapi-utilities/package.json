{
  "name": "@useoptic/openapi-utilities",
  "license": "MIT",
  "packageManager": "yarn@4.1.0",
  "version": "0.54.3",
  "main": "build/index.js",
  "types": "build/index.d.ts",
  "files": [
    "/build",
    "/src"
  ],
  "scripts": {
    "build": "yarn tsc --build --verbose",
    "test": "ENVIRONMENT=test jest --colors",
    "clean": "rm -rf build"
  },
  "devDependencies": {
    "@babel/core": "^7.17.0",
    "@babel/plugin-transform-runtime": "^7.17.0",
    "@babel/preset-env": "^7.17.0",
    "@babel/preset-typescript": "^7.17.0",
    "@types/babel__core": "^7",
    "@types/babel__preset-env": "^7",
    "@types/is-url": "^1.2.30",
    "@types/js-yaml": "^4.0.5",
    "@types/json-stable-stringify": "^1.0.33",
    "@types/lodash.groupby": "^4.6.6",
    "@types/lodash.isequal": "^4.5.6",
    "@types/lodash.omit": "^4.5.7",
    "@types/node": "^18.0.0",
    "babel-jest": "^29.3.1",
    "jest": "^29.3.1",
    "prettier": "^3.0.0",
    "ts-jest": "^29.0.3",
    "ts-node": "^10.9.1",
    "typescript": "^5.0.0"
  },
  "dependencies": {
    "@useoptic/json-pointer-helpers": "workspace:*",
    "ajv": "^8.6.0",
    "ajv-errors": "~3.0.0",
    "ajv-formats": "~2.1.0",
    "chalk": "^4.1.2",
    "fast-deep-equal": "^3.1.3",
    "is-url": "^1.2.4",
    "js-yaml": "^4.1.0",
    "json-stable-stringify": "^1.0.1",
    "lodash.groupby": "^4.6.0",
    "lodash.isequal": "^4.5.0",
    "lodash.omit": "^4.5.0",
    "node-machine-id": "^1.1.12",
    "openapi-types": "^12.0.2",
    "ts-invariant": "^0.9.3",
    "url-join": "^4.0.1",
    "yaml-ast-parser": "^0.0.43"
  }
}
