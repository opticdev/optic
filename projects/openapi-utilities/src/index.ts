import {
  OpenAPITraverser,
  normalizeOpenApiPath,
} from './openapi3/implementations/openapi3/openapi-traverser';
import { OpenAPIV2, OpenAPIV3, OpenAPIV3_1 } from 'openapi-types';
import { factsToChangelog } from './openapi3/sdk/facts-to-changelog';
export {
  FlatOpenAPIV2,
  FlatOpenAPIV3,
  FlatOpenAPIV3_1,
} from './flat-openapi-types';

import {
  ConceptualLocation,
  IChange,
  IFact,
  ILocation,
  OpenApiFact,
  OpenApiFieldFact,
  OpenApiBodyFact,
  OpenApiBodyExampleFact,
  OpenApiHeaderFact,
  OpenApiKind,
  OpenApiParameterKind,
  OpenApiOperationFact,
  OpenApiRequestParameterFact,
  OpenApiResponseFact,
  OpenApiComponentSchemaExampleFact,
  OperationLocation,
  QueryParameterLocation,
  PathParameterLocation,
  HeaderParameterLocation,
  CookieParameterLocation,
  ResponseHeaderLocation,
  ResponseLocation,
  BodyLocation,
  BodyExampleLocation,
  ComponentSchemaLocation,
  FieldLocation,
  ChangeType,
  OpenApiSpecificationFact,
  OpenApi3SchemaFact,
  FactVariant,
  ChangeVariant,
  OpenApiKindToFact,
  OpenApiRequestFact,
} from './openapi3/sdk/types';
import {
  isFactVariant,
  isChangeVariant,
  isFactOrChangeVariant,
} from './openapi3/sdk/isType';
import {
  sourcemapReader,
  getSourcemapLink,
  GetSourcemapOptions,
} from './openapi3/implementations/openapi3/sourcemap-reader';

export { defaultEmptySpec } from './openapi3/constants';
export * from './openapi3/implementations/openapi3/types';
export * from './utilities/id';

export { SPEC_TAG_REGEXP, sanitizeGitTag } from './specs/tags';

export {
  getFactForJsonPath,
  constructFactTree,
  FactTree,
} from './openapi3/json-path-interpreters';

export { UserError } from './errors';
export { getEndpointsChanges } from './utilities/changelog';
export {
  BodyChange,
  ContentType,
  OpenApiEndpointChange,
  StatusCode,
  groupChangesAndRules,
} from './utilities/group-changes';
export { traverseSpec } from './utilities/traverse-spec';
export { compareChangesByPath } from './utilities/compare-changes-by-path';
export {
  getLabel,
  getOperationsChanged,
  getOperationsChangedLabel,
} from './utilities/count-changed-operations';
export * from './swagger2';
export {
  compareSpecs,
  CompareSpecResults,
} from './compare-specs/compare-specs';
export { groupDiffsByEndpoint } from './openapi3/group-diff';
export { ObjectDiff, typeofDiff, diff, reconcileDiff } from './diff/diff';
export {
  RuleResult,
  Severity,
  textToSev,
  sevToText,
  SeverityTextOptions,
  SeverityText,
} from './results';
export { OpenApiV3Traverser, OAS3 } from './openapi3/traverser';
export { isTruthyStringValue } from './utilities/truthy';
export {
  sourcemapReader,
  GetSourcemapOptions,
  getSourcemapLink,
  OpenApiFact,
  OpenAPITraverser,
  factsToChangelog,
  ConceptualLocation,
  IChange,
  OpenApiFieldFact,
  OpenApiBodyFact,
  OpenApiBodyExampleFact,
  OpenAPIV2,
  OpenAPIV3,
  OpenAPIV3_1,
  OpenApiKind,
  OpenApiParameterKind,
  OpenApiOperationFact,
  OpenApiHeaderFact,
  IFact,
  ILocation,
  OpenApiSpecificationFact,
  OpenApiRequestParameterFact,
  OpenApiResponseFact,
  OpenApiComponentSchemaExampleFact,
  OpenApi3SchemaFact,
  OperationLocation,
  QueryParameterLocation,
  PathParameterLocation,
  HeaderParameterLocation,
  CookieParameterLocation,
  ResponseHeaderLocation,
  ResponseLocation,
  BodyLocation,
  BodyExampleLocation,
  ComponentSchemaLocation,
  FieldLocation,
  ChangeType,
  FactVariant,
  ChangeVariant,
  OpenApiKindToFact,
  isFactVariant,
  isChangeVariant,
  isFactOrChangeVariant,
  OpenApiRequestFact,
};
export { normalizeOpenApiPath };
export * from './coverage/coverage';

export * from './types';

const packageJson = require('../package.json');
export const version = packageJson.version;
