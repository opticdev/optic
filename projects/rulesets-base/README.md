# Rulesets base

Optic CI lets you write tests that enforce your API standards, and tests about how the API can change (ie breaking change rules, deprecation policy, versioning strategy, etc).

This package contains tools to write your own rules.

```javascript
import {
  Ruleset,
  OperationRule,
  RequestRule,
  ResponseRule,
  ResponseBodyRule,
  SpecificationRule,
} from '@useoptic/ruleset-base';
// or
const {
  Ruleset,
  OperationRule,
  RequestRule,
  ResponseRule,
  ResponseBodyRule,
  SpecificationRule,
} = require('@useoptic/ruleset-base');
```

## Writing your own rules

Each API Standard is expressed as a `Ruleset`. Rulesets match parts of your OpenAPI specification, and apply a set of rules to the matches.

```javascript
const postEndpointsRuleset = new Ruleset({
  name: 'POST operations standards',
  // A link to your API standards (will direct users here when rule fails)
  docsLink: 'https://optic.com/standards/post-operations',
  // The rules only run on post
  matches: (context) => context.operation.method === 'post',
  rules: [
    /* coming next */
  ],
});
```

> ðŸ’¡ More complex matches are possible. You could check for x-extensions, a certain path pattern, or anything else in the OpenAPI spec that would qualify this ruleset

Adding a rule that requires all our POST operations to require a 201 response

```javascript
const has201StatusCode = new OperationRule({
  name: 'Has 201 status codes',
  docsLink: 'https://optic.com/standards/post-operations#statuscode201',
  rule: (operationAssertions) => {
    // You can use the helper
    operationAssertions.requirement.hasStatusCodes([201]);
    // Or you can write this manually
    operationAssertions.requirement('must have 201 status code', (value) => {
      if (!value.responses.get('201'))) {
        throw new RuleError({
          message: 'Operation did not have response with status code 201',
        });
      }
    });
  },
});
```

Notice a few things

- There are helpers available for many of the rules you may want to write
- We did not provide a matches to this rule. By default it will run on any operations our Ruleset matches. Adding a matches to this rule would further scope it to a specific part of the API spec.

This rule can now be added to the above ruleset

```javascript
const postEndpointsRuleset = new Ruleset({
  name: 'POST operations standards',
  // A link to your API standards (will direct users here when rule fails)
  docsLink: 'https://optic.com/standards/post-operations',
	// The rules only run on post
  matches: (context) => context.operation.method === 'post'
  rules: [
    // This can also be shared across different Rulesets / standards
    has201StatusCode,
  ],
});
```

### Organizing your own Rulesets

When writing Rulesets for your own API Standards:

1. Group your teamâ€™s API Standards into several `Rulesets`. By HTTP Method is usually a good place to start
2. Write down the API Standards that apply to every operation, across all the groups from step 1. Usually these are things like required headers, content types, breaking change rules, etc.
3. Write code to define all your the `Rulesets` and the matchers that will qualify them `matches`
4. Start writing Rules, and add them to the `rules` property of the `Rulesets` where they apply.

### Controling where rules run

There are certain kinds of rules you want to run everywhere. Common examples include:

- Naming rules (ie snake_case)
- Breaking change rules
- Rules that help enforce your versioning strategy

By leaving `matches` unspecified, you are saying that these rules should be applied everywhere

```javascript
const namingRuleset = new Ruleset({
  name: 'Consistent Naming Standards',
  // A link to your API standards (will direct users here when rule fails)
  docsLink: 'https://optic.com/standards/naming',
  rules: [
    headersMustBeParamCase,
    queryParametersMustBeParamCase,
    requestBodyPropertiesMustBeSnakeCase,
    //...
  ],
});
```

In other cases, breaking change rules, versioning standards, and deprecation policies are triggered from changes between two versions of OpenAPI specifications. The lifecycle rules that are available are:

- `added`
- `changed`
- `removed`

```javascript
const preventResponsePropertyRemovals = new ResponseRule({
  name: 'prevent required response property removals',
  rule: (responseAssertions, context) => {
    // only runs rule on removed properties
    responseAssertions.property.removed((value) => {
      // if the property that was removed is required, throw error
      if (value.required) {
        throw new RuleError({
          message: `Must not remove required response property ${value.name}`,
        });
      }
    });
  },
});
```

### Assertion helpers

Often, there are common cases you might want to write rules for, such as "operation has query parameter" or "response body has a certain shape". Optic includes helpers to write these assertions more easily.

Examples for

```javascript
new OperationRule({
  ...,
  rule: (operationAssertions) => {
    // On operation change, checks whether the operation has a request(s) with content type
    operationAssertions.changed.hasRequests([
      { contentType: 'application/json' },
    ]);

    // On operation added, checks whether the operation has a response(s) with content type + status code
    operationAssertions.added.hasResponses([
      { statusCode: '200' }, // This checks for just the status code
      { statusCode: '400', contentType: 'application/json' }, // This expects a status code _with_ content-type
    ]);

    // Looks for a parameter that matches the shape specified
    // matches against the raw value from the OpenAPI specification
    operationAssertions.requirement.hasHeaderParameterMatching({
      name: 'X-Authorization',
    });
    operationAssertions.requirement.hasQueryParameterMatching({
      description: Matchers.string, // Looks for a description that has a string
    });
    operationAssertions.requirement.hasPathParameterMatching({
      description: Matchers.string,
      name: 'userId'
    });
  },
});
```

Request helpers

```javascript
new RequestRule({
  ...,
  rule: (requestAssertions) => {
    // Expects a partial match that the request is an object with an id: string
    requestAssertions.body.added.matches({
      schema: {
        type: 'object',
        properties: {
          id: {
            type: 'string',
          },
        },
      },
    });
  },
});
```

Response helpers

```javascript
const ruleRunner = new RuleRunner([
  new ResponseRule({
    ...,
    rule: (responseAssertions) => {
      // Matches a response header with name `X-Application' with a description type string
      responseAssertions.requirement.hasResponseHeaderMatching('X-Application', {
        description: Matchers.string,
      });
    },
  }),
]);

new ResponseBodyRule({
  ...,
  rule: (responseBodyAssertions) => {
    // Expects a partial match that the response body is an object with an id: string
    responseBodyAssertions.body.added.matches({
      schema: {
        type: 'object',
        properties: {
          id: {
            type: 'string',
          },
        },
      },
    });
  },
});
```

## Examples

[Breaking Changes](../standard-rulesets//src/breaking-changes/) and [Naming rules](../standard-rulesets//src/naming-changes/) are examples of rulesets implemented using rulesets-base.

## Testing rulesets

rulesets-base also includes a few testing helpers for writing your own tests.

```javascript
import { TestHelpers } from '@useoptic/rulesets-base';
// or
const { TestHelpers } = require('@useoptic/rulesets-base')

...
const RuleToTest = new OperationRule(...);

test('test that my rule works', () => {
  // Create some mocked data
  const beforeApiSpec = {
    ...TestHelpers.createEmptySpec(),
    paths: {
      '/api/users': {
        get: {
          responses: {}
        }
      }
    }
  };
    const afterApiSpec = {
    ...TestHelpers.createEmptySpec(),
    paths: {
      '/api/users': {
        get: {
          responses: {}
        }
      }
    }
  };
  const ruleResults = TestHelpers.runRulesWithInputs(
    [RuleToTest],
    beforeApiSpec,
    afterApiSpec
  );
  // You can use your own test library
  // e.g. with jest
  expect(ruleResults).toMatchSnapshot();
})
```
