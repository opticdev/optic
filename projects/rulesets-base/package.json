{
  "name": "@useoptic/rulesets-base",
  "license": "MIT",
  "packageManager": "yarn@4.0.2",
  "version": "0.53.20",
  "main": "build/index.js",
  "types": "build/index.d.ts",
  "files": [
    "/build",
    "/src"
  ],
  "bin": "build/index.js",
  "scripts": {
    "build": "yarn tsc --build --verbose",
    "clean": "rm -rf build",
    "test": "jest --colors"
  },
  "dependencies": {
    "@stoplight/spectral-core": "^1.8.1",
    "@stoplight/spectral-rulesets": "^1.14.1",
    "@useoptic/json-pointer-helpers": "workspace:*",
    "@useoptic/openapi-utilities": "workspace:*",
    "ajv": "^8.6.0",
    "lodash.pick": "^4.4.0",
    "node-fetch": "^2.6.7",
    "semver": "^7.5.4"
  },
  "devDependencies": {
    "@babel/core": "^7.17.0",
    "@babel/plugin-transform-runtime": "^7.17.0",
    "@babel/preset-env": "^7.17.0",
    "@babel/preset-typescript": "^7.17.0",
    "@types/babel__core": "^7",
    "@types/babel__preset-env": "^7",
    "@types/json-stable-stringify": "^1.0.33",
    "@types/lodash.pick": "^4",
    "@types/node": "^18.0.0",
    "babel-jest": "^29.3.1",
    "jest": "^29.3.1",
    "ts-jest": "^29.0.3",
    "ts-node": "^10.9.1",
    "typescript": "^5.0.0"
  }
}
