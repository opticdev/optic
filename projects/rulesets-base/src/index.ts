import * as TestHelpers from './test-helpers';
export { prepareRulesets } from './custom-rulesets/prepare-rulesets';
export { loadRuleset } from './custom-rulesets/load-ruleset';

export * from './errors';
export * from './rules';
export * from './rule-runner';
export { SpectralRule } from './extended-rules/spectral-rule';
export { TestHelpers };

export * from './types';
export type { SpectralResult, Spectral } from './extended-rules/spectral-rule';
