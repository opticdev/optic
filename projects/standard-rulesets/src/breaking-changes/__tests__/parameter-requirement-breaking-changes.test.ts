import { test, expect, describe } from '@jest/globals';
import { OpenAPIV3 } from '@useoptic/openapi-utilities';
import { TestHelpers } from '@useoptic/rulesets-base';
import { BreakingChangesRuleset } from '../index';

describe('breaking changes ruleset - parameter requirement change', () => {
  test.each(['query', 'cookie', 'header'])(
    'required %p parameter added',
    async (location: string) => {
      const beforeJson: OpenAPIV3.Document = {
        ...TestHelpers.createEmptySpec(),
        paths: {
          '/api/users': {
            get: {
              responses: {},
            },
          },
        },
      };
      const afterJson: OpenAPIV3.Document = {
        ...TestHelpers.createEmptySpec(),
        paths: {
          '/api/users': {
            get: {
              parameters: [
                {
                  name: 'required',
                  in: location,
                  required: true,
                  schema: {
                    type: 'string',
                  },
                },
              ],
              responses: {},
            },
          },
        },
      };
      const results = await TestHelpers.runRulesWithInputs(
        [new BreakingChangesRuleset()],
        beforeJson,
        afterJson
      );
      expect(results.length > 0).toBe(true);

      expect(results).toMatchSnapshot();
      expect(results.some((result) => !result.passed)).toBe(true);
    }
  );

  test.each(['query', 'cookie', 'header'])(
    '%p parameter optional to required',
    async (location: string) => {
      const beforeJson: OpenAPIV3.Document = {
        ...TestHelpers.createEmptySpec(),
        paths: {
          '/api/users': {
            get: {
              parameters: [
                {
                  name: 'version',
                  in: location,
                  schema: {
                    type: 'string',
                  },
                },
              ],
              responses: {},
            },
          },
        },
      };
      const afterJson: OpenAPIV3.Document = {
        ...TestHelpers.createEmptySpec(),
        paths: {
          '/api/users': {
            get: {
              parameters: [
                {
                  name: 'version',
                  in: location,
                  required: true,
                  schema: {
                    type: 'string',
                  },
                },
              ],
              responses: {},
            },
          },
        },
      };
      const results = await TestHelpers.runRulesWithInputs(
        [new BreakingChangesRuleset()],
        beforeJson,
        afterJson
      );
      expect(results.length > 0).toBe(true);

      expect(results).toMatchSnapshot();
      expect(results.some((result) => !result.passed)).toBe(true);
    }
  );
});
