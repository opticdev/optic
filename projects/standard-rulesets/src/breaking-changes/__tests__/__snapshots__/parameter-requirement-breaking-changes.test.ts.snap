// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`breaking changes ruleset - parameter requirement change "cookie" parameter optional to required 1`] = `
[
  {
    "change": {
      "changeType": "changed",
      "changed": {
        "after": {
          "in": "cookie",
          "name": "version",
          "required": true,
          "schema": {
            "type": "string",
          },
        },
        "before": {
          "in": "cookie",
          "name": "version",
          "schema": {
            "type": "string",
          },
        },
      },
      "location": {
        "conceptualLocation": {
          "inRequest": {
            "cookie": "version",
          },
          "method": "get",
          "path": "/api/users",
        },
        "conceptualPath": [
          "operations",
          "/api/users",
          "get",
          "parameters",
          "cookie",
          "version",
        ],
        "jsonPath": "/paths/~1api~1users/get/parameters/0",
        "kind": "cookie-parameter",
      },
    },
    "condition": undefined,
    "docsLink": undefined,
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "prevent cookie parameters enum breaking changes",
    "passed": true,
    "received": undefined,
    "severity": 2,
    "type": "changed",
    "where": "GET /api/users cookie parameter: version",
  },
  {
    "change": {
      "changeType": "changed",
      "changed": {
        "after": {
          "in": "cookie",
          "name": "version",
          "required": true,
          "schema": {
            "type": "string",
          },
        },
        "before": {
          "in": "cookie",
          "name": "version",
          "schema": {
            "type": "string",
          },
        },
      },
      "location": {
        "conceptualLocation": {
          "inRequest": {
            "cookie": "version",
          },
          "method": "get",
          "path": "/api/users",
        },
        "conceptualPath": [
          "operations",
          "/api/users",
          "get",
          "parameters",
          "cookie",
          "version",
        ],
        "jsonPath": "/paths/~1api~1users/get/parameters/0",
        "kind": "cookie-parameter",
      },
    },
    "condition": undefined,
    "docsLink": undefined,
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "prevent cookie parameters type changes",
    "passed": true,
    "received": undefined,
    "severity": 2,
    "type": "changed",
    "where": "GET /api/users cookie parameter: version",
  },
  {
    "change": {
      "changeType": "changed",
      "changed": {
        "after": {
          "in": "cookie",
          "name": "version",
          "required": true,
          "schema": {
            "type": "string",
          },
        },
        "before": {
          "in": "cookie",
          "name": "version",
          "schema": {
            "type": "string",
          },
        },
      },
      "location": {
        "conceptualLocation": {
          "inRequest": {
            "cookie": "version",
          },
          "method": "get",
          "path": "/api/users",
        },
        "conceptualPath": [
          "operations",
          "/api/users",
          "get",
          "parameters",
          "cookie",
          "version",
        ],
        "jsonPath": "/paths/~1api~1users/get/parameters/0",
        "kind": "cookie-parameter",
      },
    },
    "condition": undefined,
    "docsLink": undefined,
    "error": "cannot make optional cookie parameter 'version' required. This is a breaking change.",
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "prevent requiring existing cookie parameters",
    "passed": false,
    "received": undefined,
    "severity": 2,
    "type": "changed",
    "where": "GET /api/users cookie parameter: version",
  },
]
`;

exports[`breaking changes ruleset - parameter requirement change "header" parameter optional to required 1`] = `
[
  {
    "change": {
      "changeType": "changed",
      "changed": {
        "after": {
          "in": "header",
          "name": "version",
          "required": true,
          "schema": {
            "type": "string",
          },
        },
        "before": {
          "in": "header",
          "name": "version",
          "schema": {
            "type": "string",
          },
        },
      },
      "location": {
        "conceptualLocation": {
          "inRequest": {
            "header": "version",
          },
          "method": "get",
          "path": "/api/users",
        },
        "conceptualPath": [
          "operations",
          "/api/users",
          "get",
          "parameters",
          "header",
          "version",
        ],
        "jsonPath": "/paths/~1api~1users/get/parameters/0",
        "kind": "header-parameter",
      },
    },
    "condition": undefined,
    "docsLink": undefined,
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "prevent header parameters enum breaking changes",
    "passed": true,
    "received": undefined,
    "severity": 2,
    "type": "changed",
    "where": "GET /api/users header parameter: version",
  },
  {
    "change": {
      "changeType": "changed",
      "changed": {
        "after": {
          "in": "header",
          "name": "version",
          "required": true,
          "schema": {
            "type": "string",
          },
        },
        "before": {
          "in": "header",
          "name": "version",
          "schema": {
            "type": "string",
          },
        },
      },
      "location": {
        "conceptualLocation": {
          "inRequest": {
            "header": "version",
          },
          "method": "get",
          "path": "/api/users",
        },
        "conceptualPath": [
          "operations",
          "/api/users",
          "get",
          "parameters",
          "header",
          "version",
        ],
        "jsonPath": "/paths/~1api~1users/get/parameters/0",
        "kind": "header-parameter",
      },
    },
    "condition": undefined,
    "docsLink": undefined,
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "prevent header parameters type changes",
    "passed": true,
    "received": undefined,
    "severity": 2,
    "type": "changed",
    "where": "GET /api/users header parameter: version",
  },
  {
    "change": {
      "changeType": "changed",
      "changed": {
        "after": {
          "in": "header",
          "name": "version",
          "required": true,
          "schema": {
            "type": "string",
          },
        },
        "before": {
          "in": "header",
          "name": "version",
          "schema": {
            "type": "string",
          },
        },
      },
      "location": {
        "conceptualLocation": {
          "inRequest": {
            "header": "version",
          },
          "method": "get",
          "path": "/api/users",
        },
        "conceptualPath": [
          "operations",
          "/api/users",
          "get",
          "parameters",
          "header",
          "version",
        ],
        "jsonPath": "/paths/~1api~1users/get/parameters/0",
        "kind": "header-parameter",
      },
    },
    "condition": undefined,
    "docsLink": undefined,
    "error": "cannot make optional header parameter 'version' required. This is a breaking change.",
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "prevent requiring existing header parameters",
    "passed": false,
    "received": undefined,
    "severity": 2,
    "type": "changed",
    "where": "GET /api/users header parameter: version",
  },
]
`;

exports[`breaking changes ruleset - parameter requirement change "query" parameter optional to required 1`] = `
[
  {
    "change": {
      "changeType": "changed",
      "changed": {
        "after": {
          "in": "query",
          "name": "version",
          "required": true,
          "schema": {
            "type": "string",
          },
        },
        "before": {
          "in": "query",
          "name": "version",
          "schema": {
            "type": "string",
          },
        },
      },
      "location": {
        "conceptualLocation": {
          "inRequest": {
            "query": "version",
          },
          "method": "get",
          "path": "/api/users",
        },
        "conceptualPath": [
          "operations",
          "/api/users",
          "get",
          "parameters",
          "query",
          "version",
        ],
        "jsonPath": "/paths/~1api~1users/get/parameters/0",
        "kind": "query-parameter",
      },
    },
    "condition": undefined,
    "docsLink": undefined,
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "prevent query parameters enum breaking changes",
    "passed": true,
    "received": undefined,
    "severity": 2,
    "type": "changed",
    "where": "GET /api/users query parameter: version",
  },
  {
    "change": {
      "changeType": "changed",
      "changed": {
        "after": {
          "in": "query",
          "name": "version",
          "required": true,
          "schema": {
            "type": "string",
          },
        },
        "before": {
          "in": "query",
          "name": "version",
          "schema": {
            "type": "string",
          },
        },
      },
      "location": {
        "conceptualLocation": {
          "inRequest": {
            "query": "version",
          },
          "method": "get",
          "path": "/api/users",
        },
        "conceptualPath": [
          "operations",
          "/api/users",
          "get",
          "parameters",
          "query",
          "version",
        ],
        "jsonPath": "/paths/~1api~1users/get/parameters/0",
        "kind": "query-parameter",
      },
    },
    "condition": undefined,
    "docsLink": undefined,
    "error": undefined,
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "prevent query parameters type changes",
    "passed": true,
    "received": undefined,
    "severity": 2,
    "type": "changed",
    "where": "GET /api/users query parameter: version",
  },
  {
    "change": {
      "changeType": "changed",
      "changed": {
        "after": {
          "in": "query",
          "name": "version",
          "required": true,
          "schema": {
            "type": "string",
          },
        },
        "before": {
          "in": "query",
          "name": "version",
          "schema": {
            "type": "string",
          },
        },
      },
      "location": {
        "conceptualLocation": {
          "inRequest": {
            "query": "version",
          },
          "method": "get",
          "path": "/api/users",
        },
        "conceptualPath": [
          "operations",
          "/api/users",
          "get",
          "parameters",
          "query",
          "version",
        ],
        "jsonPath": "/paths/~1api~1users/get/parameters/0",
        "kind": "query-parameter",
      },
    },
    "condition": undefined,
    "docsLink": undefined,
    "error": "cannot make optional query parameter 'version' required. This is a breaking change.",
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "prevent requiring existing query parameters",
    "passed": false,
    "received": undefined,
    "severity": 2,
    "type": "changed",
    "where": "GET /api/users query parameter: version",
  },
]
`;

exports[`breaking changes ruleset - parameter requirement change required "cookie" parameter added 1`] = `
[
  {
    "change": {
      "added": {
        "in": "cookie",
        "name": "required",
        "required": true,
        "schema": {
          "type": "string",
        },
      },
      "changeType": "added",
      "location": {
        "conceptualLocation": {
          "inRequest": {
            "cookie": "required",
          },
          "method": "get",
          "path": "/api/users",
        },
        "conceptualPath": [
          "operations",
          "/api/users",
          "get",
          "parameters",
          "cookie",
          "required",
        ],
        "jsonPath": "/paths/~1api~1users/get/parameters/0",
        "kind": "cookie-parameter",
      },
    },
    "condition": undefined,
    "docsLink": undefined,
    "error": "cannot add required cookie parameter required to an existing operation. This is a breaking change.",
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "prevent new required cookie parameters",
    "passed": false,
    "received": undefined,
    "severity": 2,
    "type": "added",
    "where": "GET /api/users cookie parameter: required",
  },
]
`;

exports[`breaking changes ruleset - parameter requirement change required "header" parameter added 1`] = `
[
  {
    "change": {
      "added": {
        "in": "header",
        "name": "required",
        "required": true,
        "schema": {
          "type": "string",
        },
      },
      "changeType": "added",
      "location": {
        "conceptualLocation": {
          "inRequest": {
            "header": "required",
          },
          "method": "get",
          "path": "/api/users",
        },
        "conceptualPath": [
          "operations",
          "/api/users",
          "get",
          "parameters",
          "header",
          "required",
        ],
        "jsonPath": "/paths/~1api~1users/get/parameters/0",
        "kind": "header-parameter",
      },
    },
    "condition": undefined,
    "docsLink": undefined,
    "error": "cannot add required header parameter required to an existing operation. This is a breaking change.",
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "prevent new required header parameters",
    "passed": false,
    "received": undefined,
    "severity": 2,
    "type": "added",
    "where": "GET /api/users header parameter: required",
  },
]
`;

exports[`breaking changes ruleset - parameter requirement change required "query" parameter added 1`] = `
[
  {
    "change": {
      "added": {
        "in": "query",
        "name": "required",
        "required": true,
        "schema": {
          "type": "string",
        },
      },
      "changeType": "added",
      "location": {
        "conceptualLocation": {
          "inRequest": {
            "query": "required",
          },
          "method": "get",
          "path": "/api/users",
        },
        "conceptualPath": [
          "operations",
          "/api/users",
          "get",
          "parameters",
          "query",
          "required",
        ],
        "jsonPath": "/paths/~1api~1users/get/parameters/0",
        "kind": "query-parameter",
      },
    },
    "condition": undefined,
    "docsLink": undefined,
    "error": "cannot add required query parameter required to an existing operation. This is a breaking change.",
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "prevent new required query parameters",
    "passed": false,
    "received": undefined,
    "severity": 2,
    "type": "added",
    "where": "GET /api/users query parameter: required",
  },
]
`;
